File: firebase/functions/lib/events/live_meetings/get_meeting_chat_suggestion_data.dart
import 'dart:async';

import 'package:firebase_functions_interop/firebase_functions_interop.dart'
    as functions_interop;
import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import '../../on_call_function.dart';
import '../../utils/infra/firebase_auth_utils.dart';
import '../../utils/infra/firestore_utils.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/user_input/chat.dart';
import 'package:data_models/user_input/chat_suggestion_data.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/events/live_meetings/meeting_guide.dart';
import 'package:data_models/community/membership.dart';
import 'package:data_models/user/public_user_info.dart';

const bool kDebugMode = bool.fromEnvironment('dart.vm.product') == false;

class GetMeetingChatSuggestionData
    extends OnCallMethod<GetMeetingChatsSuggestionsDataRequest> {
  GetMeetingChatSuggestionData()
      : super(
          'GetMeetingChatSuggestionData',
          (jsonMap) => GetMeetingChatsSuggestionsDataRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetMeetingChatsSuggestionsDataRequest request,
    functions_interop.CallableContext context,
  ) async {
    final eventPath = request.eventPath;
    final match = RegExp('/?community/([^/]+)/templates/([^/]+)/events/([^/]+)')
        .matchAsPrefix(eventPath);
    if (match == null) {
      throw functions_interop.HttpsError(
        functions_interop.HttpsError.invalidArgument,
        'Path malformed.',
        null,
      );
    }
    final communityId = match.group(1);
    final eventId = match.group(3);

    final event = await firestoreUtils.getFirestoreObject(
      path: eventPath,
      constructor: (map) => Event.fromJson(map),
    );

    final membershipDoc =
        'memberships/${context.authUid}/community-membership/$communityId';
    final communityMembershipDoc =
        await firestore.document(membershipDoc).get();
    final membership = Membership.fromJson(
      firestoreUtils.fromFirestoreJson(communityMembershipDoc.data.toMap()),
    );

    orElseUnauthorized(
      membership.isAdmin || event.creatorId == context.authUid,
    );

    final liveMeetingPath = '${request.eventPath}/live-meetings/$eventId';
    final breakoutRoomSessions = '$liveMeetingPath/breakout-room-sessions';
    if (kDebugMode) {
      print(breakoutRoomSessions);
    }
    final sessionDocs = await firestore.collection(breakoutRoomSessions).get();
    final roomDocQueries = await Future.wait(
      sessionDocs.documents.map((session) {
        final collectionPath = '${session.reference.path}/breakout-rooms';
        if (kDebugMode) {
          print(collectionPath);
        }
        return firestore.collection(collectionPath).get();
      }),
    );
    final roomDocs =
        roomDocQueries.map((query) => query.documents).expand((a) => a);
    final breakoutMeetingLinks = roomDocs.map(
      (roomDoc) =>
          '${roomDoc.reference.path}/live-meetings/${roomDoc.documentID}',
    );

    final meetingPaths = [
      request.eventPath,
      ...breakoutMeetingLinks,
    ];

    final chatDataListResults =
        await Future.wait(<Future<List<ChatSuggestionData>>>[
      for (final path in meetingPaths) _getChatsFromPath(path),
    ]);
    final suggestionDataListResults =
        await Future.wait(<Future<List<ChatSuggestionData>>>[
      for (final path in meetingPaths) _getSuggestionsFromPath(path),
    ]);

    final agendaItemSuggestions = await _getAgendaItemSuggestions(event, [
      liveMeetingPath,
      ...breakoutMeetingLinks,
    ]);

    return GetMeetingChatsSuggestionsDataResponse(
      chatsSuggestionsList: [
        ...chatDataListResults.expand((c) => c),
        ...suggestionDataListResults.expand((c) => c),
        ...agendaItemSuggestions,
      ],
    ).toJson();
  }

  Future<List<ChatSuggestionData>> _getChatsFromPath(String path) async {
    final roomId = path.split('/').last;
    final chatsData = await firestore
        .collection('$path/chats/community_chat/messages')
        .orderBy(ChatMessage.kFieldCreatedDate)
        .get();

    final chatSuggestions = <ChatSuggestionData>[];
    for (var document in chatsData.documents) {
      var doc = ChatMessage.fromJson(
        firestoreUtils.fromFirestoreJson(document.data.toMap()),
      );
      final memberInfo = await firebaseAuthUtils.getUser(doc.creatorId!);

      String memberName = '';
      final memberDoc =
          await firestore.document('publicUser/${doc.creatorId}').get();
      if (memberDoc.exists) {
        final publicUserInfo = PublicUserInfo.fromJson(
          firestoreUtils.fromFirestoreJson(memberDoc.data.toMap()),
        );
        memberName = publicUserInfo.displayName ?? '';
      }

      chatSuggestions.add(
        ChatSuggestionData(
          id: document.documentID,
          creatorId: doc.creatorId,
          creatorEmail: memberInfo.email,
          creatorName: memberName,
          createdDate: doc.createdDate,
          message: doc.message,
          emotionType: doc.emotionType,
          type: ChatSuggestionType.chat,
          roomId: roomId,
          deleted: doc.messageStatus == ChatMessageStatus.removed,
        ),
      );
    }

    return chatSuggestions;
  }

  Future<List<ChatSuggestionData>> _getSuggestionsFromPath(String path) async {
    final roomId = path.split('/').last;
    final userSuggestionsData = await firestore
        .collection('$path/user-suggestions')
        .orderBy(SuggestedAgendaItem.kFieldCreatedDate)
        .get();

    final chatSuggestionDataList = <ChatSuggestionData>[];
    for (var document in userSuggestionsData.documents) {
      var doc = SuggestedAgendaItem.fromJson(
        firestoreUtils.fromFirestoreJson(document.data.toMap()),
      );
      final memberInfo = await firebaseAuthUtils.getUser(doc.creatorId!);

      final memberDoc =
          await firestore.document('publicUser/${doc.creatorId}').get();
      String memberName = '';
      if (memberDoc.exists) {
        final publicUserInfo = PublicUserInfo.fromJson(
          firestoreUtils.fromFirestoreJson(memberDoc.data.toMap()),
        );
        memberName = publicUserInfo.displayName ?? '';
      }

      chatSuggestionDataList.add(
        ChatSuggestionData(
          id: document.documentID,
          creatorId: doc.creatorId,
          creatorEmail: memberInfo.email,
          creatorName: memberName,
          createdDate: doc.createdDate,
          message: doc.content,
          type: ChatSuggestionType.suggestion,
          upvotes: doc.upvotedUserIds.length,
          downvotes: doc.downvotedUserIds.length,
          roomId: roomId,
        ),
      );
    }

    return chatSuggestionDataList;
  }

  Future<List<ChatSuggestionData>> _getAgendaItemSuggestions(
    Event event,
    List<String> roomPaths,
  ) async {
    final suggestionAgendaItems = event.agendaItems
        .where((a) => a.type == AgendaItemType.userSuggestions)
        .map((a) => a.id)
        .toList();

    final participantDetailsCollections = roomPaths
        .map(
          (path) => suggestionAgendaItems.map(
            (suggestionAgendaItemId) =>
                '$path/participant-agenda-item-details/$suggestionAgendaItemId/participant-details',
          ),
        )
        .expand((element) => element);

    if (kDebugMode) {
      print(participantDetailsCollections);
    }
    final participantAgendaItemDetails = await Future.wait(
      participantDetailsCollections
          .map((collection) => _getParticipantAgendaItemDetails(collection)),
    );

    return participantAgendaItemDetails.expand((e) => e).toList();
  }

  Future<List<ChatSuggestionData>> _getParticipantAgendaItemDetails(
    String collection,
  ) async {
    final documents = await firestore.collection(collection).get();

    final results = await Future.wait(
      documents.documents
          .map(
            (doc) => _getSuggestionsFromParticipantAgendaItem(
              doc,
              ParticipantAgendaItemDetails.fromJson(
                firestoreUtils.fromFirestoreJson(doc.data.toMap()),
              ),
            ),
          )
          .toList(),
    );

    return results.expand((element) => element).toList();
  }

  Future<List<ChatSuggestionData>> _getSuggestionsFromParticipantAgendaItem(
    DocumentSnapshot document,
    ParticipantAgendaItemDetails details,
  ) async {
    final memberInfo = await firebaseAuthUtils.getUser(details.userId!);

    final memberDoc =
        await firestore.document('publicUser/${details.userId}').get();
    String memberName = '';
    if (memberDoc.exists) {
      final publicUserInfo = PublicUserInfo.fromJson(
        firestoreUtils.fromFirestoreJson(memberDoc.data.toMap()),
      );
      memberName = publicUserInfo.displayName ?? '';
    }

    return details.suggestions
        .map(
          (suggestion) => ChatSuggestionData(
            id: document.documentID,
            creatorId: details.userId,
            creatorEmail: memberInfo.email,
            creatorName: memberName,
            createdDate:
                suggestion.createdDate ?? document.createTime?.toDateTime(),
            message: suggestion.suggestion,
            type: ChatSuggestionType.suggestion,
            upvotes: suggestion.likedByIds.length,
            downvotes: suggestion.dislikedByIds.length,
            roomId: details.meetingId,
            agendaItemId: details.agendaItemId,
          ),
        )
        .toList();
  }
}

File: firebase/functions/src/events/live_meetings/get_meeting_chat_suggestion_data.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { firebaseAuthUtils } from '../../utils/infra/firebase_auth_utils';
import { orElseUnauthorized } from '../../utils/utils';
import { Event, Membership } from '../../types';

interface GetMeetingChatsSuggestionsDataRequest {
  eventPath: string;
}

export class GetMeetingChatSuggestionData extends OnCallMethod<GetMeetingChatsSuggestionsDataRequest> {
  constructor() {
    super(
      'GetMeetingChatSuggestionData',
      (jsonMap) => jsonMap as GetMeetingChatsSuggestionsDataRequest
    );
  }

  async action(
    request: GetMeetingChatsSuggestionsDataRequest,
    context: functions.https.CallableContext
  ): Promise<Record<string, unknown>> {
    const eventPath = request.eventPath;
    const match = /\/?community\/([^/]+)\/templates\/([^/]+)\/events\/([^/]+)/.exec(eventPath);
    if (!match) throw new functions.https.HttpsError('invalid-argument', 'Path malformed.');

    const communityId = match[1];
    const eventId = match[3];

    const event: Event = await firestoreUtils.getFirestoreObject({
      path: eventPath,
      constructor: (map) => map as unknown as Event,
    });

    const membershipSnap = await firestore
      .doc(`memberships/${context.auth?.uid}/community-membership/${communityId}`)
      .get();
    const membership = firestoreUtils.fromFirestoreJson(membershipSnap.data() ?? {}) as unknown as Membership;

    orElseUnauthorized((membership as any).isAdmin || (event as any).creatorId === context.auth?.uid);

    const liveMeetingPath = `${request.eventPath}/live-meetings/${eventId}`;
    const sessionDocs = await firestore.collection(`${liveMeetingPath}/breakout-room-sessions`).get();
    const roomDocQueries = await Promise.all(
      sessionDocs.docs.map((session) => firestore.collection(`${session.ref.path}/breakout-rooms`).get())
    );
    const roomDocs = roomDocQueries.flatMap((q) => q.docs);
    const breakoutMeetingLinks = roomDocs.map((roomDoc) => `${roomDoc.ref.path}/live-meetings/${roomDoc.id}`);
    const meetingPaths = [request.eventPath, ...breakoutMeetingLinks];

    const [chatDataListResults, suggestionDataListResults] = await Promise.all([
      Promise.all(meetingPaths.map((path) => this._getChatsFromPath(path))),
      Promise.all(meetingPaths.map((path) => this._getSuggestionsFromPath(path))),
    ]);

    const agendaItemSuggestions = await this._getAgendaItemSuggestions(event, [
      liveMeetingPath,
      ...breakoutMeetingLinks,
    ]);

    return {
      chatsSuggestionsList: [
        ...chatDataListResults.flat(),
        ...suggestionDataListResults.flat(),
        ...agendaItemSuggestions,
      ],
    };
  }

  private async _getChatsFromPath(path: string): Promise<Record<string, unknown>[]> {
    const roomId = path.split('/').pop()!;
    const chatsData = await firestore
      .collection(`${path}/chats/community_chat/messages`)
      .orderBy('createdDate')
      .get();

    const chatSuggestions: Record<string, unknown>[] = [];
    for (const document of chatsData.docs) {
      const doc = firestoreUtils.fromFirestoreJson(document.data()) as Record<string, unknown>;
      const memberInfo = await firebaseAuthUtils.getUser(doc['creatorId'] as string);
      const memberDoc = await firestore.doc(`publicUser/${doc['creatorId']}`).get();
      const memberName = (memberDoc.data() ?? {})['displayName'] as string ?? '';

      chatSuggestions.push({
        id: document.id,
        creatorId: doc['creatorId'],
        creatorEmail: (memberInfo as any)?.email,
        creatorName: memberName,
        createdDate: doc['createdDate'],
        message: doc['message'],
        emotionType: doc['emotionType'],
        type: 'chat',
        roomId,
        deleted: doc['messageStatus'] === 'removed',
      });
    }
    return chatSuggestions;
  }

  private async _getSuggestionsFromPath(path: string): Promise<Record<string, unknown>[]> {
    const roomId = path.split('/').pop()!;
    const userSuggestionsData = await firestore
      .collection(`${path}/user-suggestions`)
      .orderBy('createdDate')
      .get();

    const results: Record<string, unknown>[] = [];
    for (const document of userSuggestionsData.docs) {
      const doc = firestoreUtils.fromFirestoreJson(document.data()) as Record<string, unknown>;
      const memberInfo = await firebaseAuthUtils.getUser(doc['creatorId'] as string);
      const memberDoc = await firestore.doc(`publicUser/${doc['creatorId']}`).get();
      const memberName = (memberDoc.data() ?? {})['displayName'] as string ?? '';

      results.push({
        id: document.id,
        creatorId: doc['creatorId'],
        creatorEmail: (memberInfo as any)?.email,
        creatorName: memberName,
        createdDate: doc['createdDate'],
        message: doc['content'],
        type: 'suggestion',
        upvotes: ((doc['upvotedUserIds'] as unknown[]) ?? []).length,
        downvotes: ((doc['downvotedUserIds'] as unknown[]) ?? []).length,
        roomId,
      });
    }
    return results;
  }

  private async _getAgendaItemSuggestions(event: Event, roomPaths: string[]): Promise<Record<string, unknown>[]> {
    const suggestionAgendaItems = ((event as any).agendaItems ?? [])
      .filter((a: any) => a.type === 'userSuggestions')
      .map((a: any) => a.id as string);

    const collections: string[] = roomPaths.flatMap((path) =>
      suggestionAgendaItems.map(
        (id: string) => `${path}/participant-agenda-item-details/${id}/participant-details`
      )
    );

    const results = await Promise.all(collections.map((c) => this._getParticipantAgendaItemDetails(c)));
    return results.flat();
  }

  private async _getParticipantAgendaItemDetails(collection: string): Promise<Record<string, unknown>[]> {
    const documents = await firestore.collection(collection).get();
    const results = await Promise.all(
      documents.docs.map((doc) => {
        const details = firestoreUtils.fromFirestoreJson(doc.data()) as Record<string, unknown>;
        return this._getSuggestionsFromParticipantAgendaItem(doc.id, details);
      })
    );
    return results.flat();
  }

  private async _getSuggestionsFromParticipantAgendaItem(
    docId: string,
    details: Record<string, unknown>
  ): Promise<Record<string, unknown>[]> {
    const userId = details['userId'] as string | undefined;
    if (!userId) return [];

    const memberInfo = await firebaseAuthUtils.getUser(userId);
    const memberDoc = await firestore.doc(`publicUser/${userId}`).get();
    const memberName = (memberDoc.data() ?? {})['displayName'] as string ?? '';

    const suggestions = (details['suggestions'] as unknown[]) ?? [];
    return suggestions.map((suggestion: any) => ({
      id: docId,
      creatorId: userId,
      creatorEmail: (memberInfo as any)?.email,
      creatorName: memberName,
      createdDate: suggestion['createdDate'],
      message: suggestion['suggestion'],
      type: 'suggestion',
      upvotes: (suggestion['likedByIds'] ?? []).length,
      downvotes: (suggestion['dislikedByIds'] ?? []).length,
      roomId: details['meetingId'],
      agendaItemId: details['agendaItemId'],
    }));
  }
}
